libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -95,15 +105,31 @@ public MicrosoftAppCredentials(string appId, string password, HttpClient customH
         /// </value>
         public string MicrosoftAppPassword { get; set; }
 
+        /// <summary>
+        /// Gets or sets the channel auth token tenant for this credential.
+        /// </summary>
+        private string _channelAuthTenant;
+        public string ChannelAuthTenant
+        {
+            get { return string.IsNullOrEmpty(_channelAuthTenant) ? "botframework.com" : _channelAuthTenant; }
+            private set { _channelAuthTenant = value; }
+        }
+
         /// <summary>
         /// Gets the OAuth endpoint to use.
         /// </summary>
-        /// <value>
-        /// The OAuth endpoint to use.
-        /// </value>
         public virtual string OAuthEndpoint
         {
-            get { return AuthenticationConstants.ToChannelFromBotLoginUrl; }
+            get
+            {
+                // Advanced user only, see https://aka.ms/bots/tenant-restriction
+                var endpointUrl = string.Format(AuthenticationConstants.ToChannelFromBotLoginUrlTemplate, ChannelAuthTenant);
+
+                if (Uri.TryCreate(endpointUrl, UriKind.Absolute, out var result))
```

Comment:
This doesn't pass stylecop. Please make sure this builds with no warnings.

Reply:
k :)